libs/execution/src/lib/types/io-types/filesystem-inmemory.ts

```diff
@@ -4,94 +4,96 @@
-      const fileName = processedParts[processedParts.length - 1] as string;
+    // Loop until we reach the last part of processedParts
+    for (let i = 0; i < processedParts.length - 1; i++) {
+      const part = processedParts[i];
 
-      // If we have a directory, we traverse it, if not (processedParts == 1) skip the travesal
-      for (let i = 0; i < processedParts.length - 1; i++) {
-        const childFileSystem = currentFileSystemIndex.get(
-          processedParts[i] as string,
-        ) as InMemoryFileSystem | undefined;
+      // If part exists
+      if (part != null) {
+        let childNode = currentDir.getChildNode(part);
 
-        // If we dont find current path-part, stop methodcall
-        if (!childFileSystem) {
+        // If dir exists --> traverse
+        if (childNode && childNode instanceof FileSystemDirectory) {
+          currentDir = childNode;
+        } else if (childNode && childNode instanceof FileSystemFile) {
           return null;
+
+          // If dir NOT exists --> create new
+        } else {
+          childNode = new FileSystemDirectory(part);
+          currentDir.add(childNode);
+          currentDir = childNode as FileSystemDirectory;
         }
-        currentFileSystemIndex = childFileSystem.fileSystemIndex;
       }
+    }
 
-      // If we are in the correct directory
-      const file = currentFileSystemIndex.get(fileName) as
-        | BinaryFile
-        | undefined;
-      return file === undefined ? null : file;
+    // If Nodename is not already there --> put file to dir
+    const child = currentDir.getChildNode(processedParts.pop() ?? '');
+    if (child instanceof FileSystemDirectory || child === undefined) {
+      currentDir.add(file);
     }
     return null;
   }
 
-  putFile(filePath: string, file: BinaryFile): FileSystem {
-    const processedParts = this.processPath(filePath);
-    let currentFileSystemIndex = this.fileSystemIndex;
-
-    // If we need to traverse
-    for (let i = 0; i < processedParts.length - 1; i++) {
-      const part: string = processedParts[i] as string;
-
-      // Check, if directory already exists, if not create it
-      let childFileSystemIndex = currentFileSystemIndex.get(part) as
-        | InMemoryFileSystem
-        | undefined;
-      if (!childFileSystemIndex) {
-        childFileSystemIndex = new InMemoryFileSystem();
-        currentFileSystemIndex.set(part, childFileSystemIndex);
+  private processPath(path: string): string[] {
+    const parts = path
+      .toLowerCase()
+      .split(InMemoryFileSystem.PATH_SEPARATOR)
+      .filter((p) => p !== ''); // Process paths like "folder1//folder1" to "folder1/folder2"
+    const processedParts: string[] = [];
+    for (const part of parts) {
+      if (part === InMemoryFileSystem.CURRENT_DIR) {
+        continue; // Skip current dirs in path
+      }
+      if (part === InMemoryFileSystem.PARENT_DIR) {
+        processedParts.pop(); // Go level up in folder hierarchy if ..
```

Comment:
What if `processedParts` is empty at this point? In your current implementation, a path `../folder/file.txt` is semantically equal to `./folder/file.txt` which should not be the case.

Reply:
Ouh thats a good point, thanks for that hint! 

Actually in the current implementation of `InMemoryFileSystem` which always starts at its "virtual" root-dir it does not make sense to call getFile/putFile in general with a relative path (like relativ to what since it is in memory?).

I would mention this in the documentation, if this is okay?

Reply:
Yes, I'd argue that you should always call it with absolute paths, so you always start in the root directory. Then you should ensure, that the method is only called with absolute paths and error on any paths that ascent beyond the root directory (e.g. `/folder/../..`).

Reply:
is implemented in new version